# Part-of-speech tagging and Named Entity Recognition

In this project, we'll demonstrate how we can perform **part-of-speech tagging (POS) and named entity recognition (NER) with spaCy**.

For this, we will use two different books, both freely available on the [Project Gutenberg website](https://www.gutenberg.org):

* "Flatland: A Romance of Many Dimensions", by Edwin Abbott Abbott
<br><br>
* Charles Darwin's seminal book "On the Origin of Species by Means of Natural Selection"

### Part-of-speech tagging

#### 1. Perform initial imports

In [1]:
import spacy

nlp = spacy.load('en_core_web_sm')

from spacy import displacy

#### 2. Load data

In [2]:
with open('./text/flatland.txt', encoding='utf8') as f:
    doc = nlp(f.read())

#### 3. Explore data

In spaCy, the `Doc` class is a container for accessing linguistic annotations. Let's explore the `doc` object we've just created via the `nlp` object.

In [3]:
# sentences of our doc

sentences = list(doc.sents)

In [4]:
# sentence example

sentences[11]

I call our world Flatland, not because we call it so, but to make its
nature clearer to you, my happy readers, who are privileged to live in
Space.


In [5]:
# total number of sentences

len(sentences)

1667

In [6]:
# span of our sentence

sentences[11][0:5]

I call our world Flatland

#### 4. Extract POS tags

Let's see how we can access the **text, POS tags, fine-grained POS tags and the description of these fine-grained POS tags**.

In [7]:
# text and POS tag

for token in sentences[11][0:5]:
    print(token.text, token.pos_)

I PRON
call VERB
our DET
world NOUN
Flatland PROPN


In [8]:
# text, POS tag, fine-grained POS tag and description for whole sentence

for token in sentences[11]:
    print(f'{token.text:{15}} {token.pos_:{8}} {token.tag_:{8}} {spacy.explain(token.tag_)}')

I               PRON     PRP      pronoun, personal
call            VERB     VBP      verb, non-3rd person singular present
our             DET      PRP$     pronoun, possessive
world           NOUN     NN       noun, singular or mass
Flatland        PROPN    NNP      noun, proper singular
,               PUNCT    ,        punctuation mark, comma
not             PART     RB       adverb
because         SCONJ    IN       conjunction, subordinating or preposition
we              PRON     PRP      pronoun, personal
call            VERB     VBP      verb, non-3rd person singular present
it              PRON     PRP      pronoun, personal
so              ADV      RB       adverb
,               PUNCT    ,        punctuation mark, comma
but             CCONJ    CC       conjunction, coordinating
to              PART     TO       infinitival "to"
make            VERB     VB       verb, base form
its             DET      PRP$     pronoun, possessive

               SPACE    _SP      None
natur

We can also easily count the total number of each POS tag in our document.

In [9]:
POS_counts = doc.count_by(spacy.attrs.POS)

POS_counts

{96: 2310,
 103: 2748,
 95: 2313,
 90: 4880,
 97: 4955,
 87: 1540,
 84: 2339,
 85: 4053,
 92: 5149,
 89: 1501,
 100: 4026,
 93: 451,
 94: 845,
 98: 968,
 86: 2003,
 91: 55,
 99: 2,
 101: 15}

Let's order them and replace their ID's by the corresponding strings.

In [10]:
for key ,value in sorted(POS_counts.items(), key=lambda item: item[1], reverse=True):
    print(f'{doc.vocab[key].text:{8}}: {value}')

NOUN    : 5149
PUNCT   : 4955
DET     : 4880
ADP     : 4053
VERB    : 4026
SPACE   : 2748
ADJ     : 2339
PRON    : 2313
PROPN   : 2310
ADV     : 2003
AUX     : 1540
CCONJ   : 1501
SCONJ   : 968
PART    : 845
NUM     : 451
INTJ    : 55
X       : 15
SYM     : 2


Nouns and punctuation marks are the most common POS tags.

#### 5. Visualize dependency parse

Based on their POS tags, we can also visualize the relationship between our words in a sentence.

In [11]:
# considering only our first 5 tokens

displacy.render(list(doc.sents)[11][:5], style='dep', options={'compact': True, 'bg': '#09a3d5', 'color': 'white'})

### Named Entity Recoginition

#### 6. Load data

In [12]:
with open('./text/origin_of_species.txt', encoding='utf8') as f:
    doc = nlp(f.read())

#### 7. Explore data

In [13]:
sentences = list(doc.sents)

In [14]:
sentences[179]

In considering the Origin of Species, it is quite conceivable that a
naturalist, reflecting on the mutual affinities of organic beings, on their
embryological relations, their geographical distribution, geological
succession, and other such facts, might come to the conclusion that each
species had not been independently created, but had descended, like
varieties, from other species.

#### 8. Recognize named entities

In [15]:
# function to display entity info

def show_ents(doc):
    if doc.ents:
        for ent in doc.ents:
            print(ent.text+': '+ent.label_+' - '+str(spacy.explain(ent.label_)))
    else:
        print('No named entities found.')

In [16]:
show_ents(sentences[179])

the Origin of Species: WORK_OF_ART - Titles of books, songs, etc.


In [17]:
sentences[180]

Nevertheless, such a conclusion, even if
well founded, would be unsatisfactory, until it could be shown how the
innumerable species inhabiting this world have been modified, so as to
acquire that perfection of structure and coadaptation which most justly
excites our admiration.

In [18]:
show_ents(sentences[180])

No named entities found.


#### 8. Identify sentences with 'n' or more named entities

In [19]:
def sents_n_ents(doc, n=1):
    list_sentences = [sentence for sentence in doc.sents if len(sentence.ents) >= n]
    return list_sentences

In [20]:
# sentences with 4 or more named entities

ner_sentences = sents_n_ents(doc, 4)

In [21]:
len(ner_sentences)

264

We have 264 sentences with 4 or more named entities.

#### 9. Recognize named entities in one of these sentences

In [22]:
ner_sentences[14]

" Pigeons were much valued by Akber Khan
in India, about the year 1600; never less than 20,000 pigeons were taken
with the court.

In [23]:
show_ents(ner_sentences[14])

Akber Khan: PERSON - People, including fictional
India: GPE - Countries, cities, states
about the year 1600: DATE - Absolute or relative dates or periods
less than 20,000: CARDINAL - Numerals that do not fall under another type


#### 10. Visualize named entities

In [24]:
displacy.render(ner_sentences[14], style='ent')

#### 11. Add a named entity

In [25]:
ner_sentences[24]

if that
between America and Europe is ample, will that between the Continent and
the Azores, or Madeira, or the Canaries, or Ireland, be sufficient?

In [26]:
show_ents(ner_sentences[24])

America: GPE - Countries, cities, states
Europe: LOC - Non-GPE locations, mountain ranges, bodies of water
Continent: LOC - Non-GPE locations, mountain ranges, bodies of water
Canaries: LOC - Non-GPE locations, mountain ranges, bodies of water
Ireland: GPE - Countries, cities, states


As we can see, [Azores](https://en.wikipedia.org/wiki/Azores) and [Madeira](https://en.wikipedia.org/wiki/Madeira) are not recognized as a named entity - LOC - by spaCy. Let's change that!

In order to do this, we need to know the position - the indices - of our tokens of interest in our `doc` object.

In [27]:
token_i_madeira = [token.i for token in doc if token.text == 'Madeira']

In [28]:
token_i_azores = [token.i for token in doc if token.text == 'Azores']

Since we know that for our sentence of interest the token "Madeira" is the third token after the token "Azores", we can use that information to identify our indices.

In [29]:
possible_i=[]

for token_i in token_i_azores:
    if token_i + 3 in token_i_madeira:
        possible_i.append(token_i)

In [30]:
possible_i

[20728]

Luckily, there's only one possibility. Let's see if it is correct.

In [31]:
doc[20728:20743]

Azores, or Madeira, or the Canaries, or Ireland, be sufficient?

It seems to be our sentence! Let's confirm our indices and add a named entity to these tokens.

In [32]:
doc[20728]

Azores

In [33]:
doc[20731]

Madeira

In [34]:
from spacy.tokens import Span

In [35]:
azores_ent = Span(doc, 20728, 20729, label='LOC')
madeira_ent = Span(doc, 20731, 20732, label='LOC')

doc.ents = list(doc.ents) + [azores_ent, madeira_ent]

Let's see if this worked.

In [36]:
show_ents(ner_sentences[24])

America: GPE - Countries, cities, states
Europe: LOC - Non-GPE locations, mountain ranges, bodies of water
Continent: LOC - Non-GPE locations, mountain ranges, bodies of water
Azores: LOC - Non-GPE locations, mountain ranges, bodies of water
Madeira: LOC - Non-GPE locations, mountain ranges, bodies of water
Canaries: LOC - Non-GPE locations, mountain ranges, bodies of water
Ireland: GPE - Countries, cities, states


In [37]:
displacy.render(ner_sentences[24], style='ent')

Everything is working as expected.

However, this means that Azores and Madeira will be identified as entities in this sentence, but not in other occurrences throughout our document. In order to change this, we'll use spaCy's **PhraseMatcher**.

#### 12. Add a named entity to all occurences of a token

In [38]:
from spacy.matcher import PhraseMatcher

In [39]:
# initialize the matcher

matcher = PhraseMatcher(nlp.vocab)

In [40]:
# create match patterns and add them to the matcher

pattern1 = nlp("Azores")
pattern2 = nlp("Madeira")

matcher.add('PATTERN1', None, pattern1)
matcher.add('PATTERN2', None, pattern2)

In [41]:
# apply the matcher to our doc

for match_id, start, end in matcher(doc):
    print(doc.vocab.strings[match_id]+': ', str(start)+' --> '+str(end), ' ('+doc[start:end].text+')')

PATTERN2:  20570 --> 20571  (Madeira)
PATTERN1:  20728 --> 20729  (Azores)
PATTERN2:  20731 --> 20732  (Madeira)
PATTERN2:  21980 --> 21981  (Madeira)
PATTERN2:  42782 --> 42783  (Madeira)
PATTERN2:  53622 --> 53623  (Madeira)
PATTERN2:  53690 --> 53691  (Madeira)
PATTERN2:  53731 --> 53732  (Madeira)
PATTERN2:  53792 --> 53793  (Madeira)
PATTERN2:  53895 --> 53896  (Madeira)
PATTERN2:  54434 --> 54435  (Madeira)
PATTERN1:  55446 --> 55447  (Azores)
PATTERN2:  120501 --> 120502  (Madeira)
PATTERN2:  129997 --> 129998  (Madeira)
PATTERN1:  138931 --> 138932  (Azores)
PATTERN2:  149084 --> 149085  (Madeira)
PATTERN2:  149475 --> 149476  (Madeira)
PATTERN2:  149517 --> 149518  (Madeira)
PATTERN2:  149608 --> 149609  (Madeira)
PATTERN2:  149716 --> 149717  (Madeira)
PATTERN2:  150302 --> 150303  (Madeira)
PATTERN1:  150305 --> 150306  (Azores)
PATTERN2:  153942 --> 153943  (Madeira)
PATTERN2:  153984 --> 153985  (Madeira)
PATTERN1:  187445 --> 187446  (Azores)
PATTERN2:  187645 --> 187646 

We have now identified all the occurences of our tokens "Azores" and "Madeira". Let's add a named entity to all of them.

Apparently, several occurrences of the token "Madeira" were identified with the named entity "ORG". Since we cannot have overlapping entities, let's try to do this only for the "Azores" token.

In [44]:
# initialize the matcher

matcher = PhraseMatcher(nlp.vocab)

In [45]:
# create match patterns and add them to the matcher

pattern = nlp("Azores")

matcher.add('PATTERN', None, pattern)

In [46]:
# apply the matcher to our doc

for match_id, start, end in matcher(doc):
    print(doc.vocab.strings[match_id]+': ', str(start)+' --> '+str(end), ' ('+doc[start:end].text+')')

PATTERN:  20728 --> 20729  (Azores)
PATTERN:  55446 --> 55447  (Azores)
PATTERN:  138931 --> 138932  (Azores)
PATTERN:  150305 --> 150306  (Azores)
PATTERN:  187445 --> 187446  (Azores)
PATTERN:  187916 --> 187917  (Azores)
PATTERN:  190384 --> 190385  (Azores)
PATTERN:  195084 --> 195085  (Azores)


In [47]:
# new entities excluding the one we already added to avoid conflicts
# there's also other conflicting doc.ents with start=187916 and 190384

new_ents = [Span(doc, start, end, label='LOC') for _, start, end in matcher(doc) 
            if (start != 20728) and (start !=187916) and (start !=190384)]

doc.ents = list(doc.ents) + new_ents

Let's check if this worked.

In [48]:
# for pattern with start=55446

doc[55411:55450]

Mr. Thwaites informs me that he
has observed similar facts in Ceylon, and analogous observations have been
made by Mr. H. C. Watson on European species of plants brought from the
Azores to England.

In [49]:
show_ents(doc[55411:55450])

Thwaites: PERSON - People, including fictional
Ceylon: GPE - Countries, cities, states
H. C. Watson: PERSON - People, including fictional
European: NORP - Nationalities or religious or political groups
Azores: LOC - Non-GPE locations, mountain ranges, bodies of water
England: GPE - Countries, cities, states


In [50]:
displacy.render(doc[55411:55450], style='ent')

As expected, the "Azores" token is now identified as a LOC - Non-GPE locations, mountain ranges, bodies of water.